bluemira/equilibria/opt_problems.py

```diff
@@ -0,0 +1,813 @@
+        )
+
+        # Set up data needed in FoM evaluation.
+        # Scale the control matrix and constraint vector by weights.
+        self.targets(self.eq, I_not_dI=True)
+        _, a_mat, b_vec = self.targets.get_weighted_arrays()
+
+        self._objective._args["scale"] = self.scale
+        self._objective._args["a_mat"] = a_mat
+        self._objective._args["b_vec"] = b_vec
+
+        # Optimise
+        currents = self.opt.optimise(initial_currents)
+
+        coilset_state = np.concatenate((self.x0, self.z0, currents))
+        self.set_coilset_state(self.coilset, coilset_state, self.scale)
+        return self.coilset
+
+
+class CoilsetPositionCOP(CoilsetOP):
+    """
+    Coilset OptimisationProblem for coil currents and positions
+    subject to maximum current bounds and positions bounded within
+    a provided region.
+
+    Coil currents and positions are optimised simultaneously.
+
+    Parameters
+    ----------
+    coilset: CoilSet
+        Coilset to optimise.
+    eq: Equilibrium
+        Equilibrium object used to update magnetic field targets.
+    targets: MagneticConstraintSet
+        Set of magnetic field targets to use in objective function.
+    pfregions: dict(coil_name:Loop, coil_name:Loop, ...)
+        Dictionary of loops that specify convex hull regions inside which
+        each PF control coil position is to be optimised.
+        The loop objects must be 2d in x,z in units of [m].
+    max_currents: float or np.array(len(coilset._ccoils)) (default = None)
+        Maximum allowed current for each independent coil current in coilset [A].
+        If specified as a float, the float will set the maximum allowed current
+        for all coils.
+    gamma: float (default = 1e-8)
+        Tikhonov regularisation parameter in units of [A⁻¹].
+    optimiser: Optimiser
+        Optimiser object to use for constrained optimisation.
+    constraints: List[OptimisationConstraint] (default: [])
+        Optional list of OptimisationConstraint objects storing
+        information about constraints that must be satisfied
+        during the coilset optimisation, to be provided to the
+        optimiser.
+
+    Notes
+    -----
+    Setting stopval and maxeval is the most reliable way to stop optimisation
+    at the desired figure of merit and number of iterations respectively.
+    Some NLOpt optimisers display unexpected behaviour when setting xtol and
+    ftol, and may not terminate as expected when those criteria are reached.
+    """
+
+    def __init__(
+        self,
+        coilset: CoilSet,
+        eq: Equilibrium,
+        targets: MagneticConstraintSet,
+        pfregions: dict,
+        max_currents=None,
+        gamma=1e-8,
+        optimiser=Optimiser(
+            algorithm_name="SBPLX",
+            opt_conditions={
+                "stop_val": 1.0,
+                "max_eval": 100,
+            },
+            opt_parameters={},
+        ),
+        opt_constraints=[],
```

Comment:
I'm a bit wary of setting mutable objects as defaults to function arguments, both here and opt_parameters just above. Can we do a `None` and internally convert to `[]` etc jump about

https://stackoverflow.com/questions/1132941/least-astonishment-and-the-mutable-default-argument

Reply:
Thanks - that could have created some real headaches - particularly for constraints!

Should be noted that the NLOptOptimiser API also suffers from this... unlikely to be too much of an issue, unless it's init'd multiple times, though.

Reply:
Probably should split that out into an issue